sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ExceptionsNeedStandardConstructors.cs

```diff
@@ -0,0 +1,87 @@
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class ExceptionsNeedStandardConstructors : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4027";
+        private const string MessageFormat = "Implement the missing constructors for this exception.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                var classDeclaration = c.Node as ClassDeclarationSyntax;
+                var classSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+
+                if (classSymbol.DerivesFrom(KnownType.System_Exception) &&
+                    !HasStandardConstructors(classSymbol) &&
+                    !classDeclaration.Identifier.IsMissing)
+                {
+                    c.ReportDiagnostic(Diagnostic.Create(rule, classDeclaration.Identifier.GetLocation()));
+                }
+            },
+            SyntaxKind.ClassDeclaration);
+        }
+
+        private static bool HasStandardConstructors(INamedTypeSymbol classSymbol)
+        {
+            var ctors = classSymbol.Constructors;
+
+            return HasConstructor(ctors, Accessibility.Public) &&
+                   HasConstructor(ctors, Accessibility.Public, KnownType.System_String) &&
+                   HasConstructor(ctors, Accessibility.Public, KnownType.System_String, KnownType.System_Exception) &&
+                   HasConstructor(ctors,
+                       classSymbol.IsSealed ? Accessibility.Private : Accessibility.Protected,
+                       KnownType.System_Runtime_Serialization_SerializationInfo,
+                       KnownType.System_Runtime_Serialization_StreamingContext);
+        }
+
+        private static bool HasConstructor(ImmutableArray<IMethodSymbol> constructors,
+            Accessibility accessibility, params KnownType[] expectedParameterTypes)
+        {
+            return constructors.Any(c => IsMatchingConstructor(c, accessibility, expectedParameterTypes));
+        }
+
+        private static bool IsMatchingConstructor(IMethodSymbol constructor, Accessibility accessibility,
+            KnownType[] expectedParameterTypes)
+        {
+            return constructor.DeclaredAccessibility == accessibility &&
+                   CollectionUtils.AreEqual(constructor.Parameters, expectedParameterTypes,
```

Comment:
It is not and I prefer it this way. I don't like how asymmetrical it is when doing a.Equals(b). Equals(a, b) reads much better.